# Combining Data

Practice combining data from two different data sets. In the same folder as this Jupyter notebook, there are two csv files:
* rural_population_percent.csv
* electricity_access_percent.csv

They both come from the World Bank Indicators data. 
* https://data.worldbank.org/indicator/SP.RUR.TOTL.ZS
* https://data.worldbank.org/indicator/EG.ELC.ACCS.ZS

The rural populaton data represents the percent of a country's population that is rural over time. The electricity access data shows the percentage of people with access to electricity.

In this exercise, you will combine these two data sets together into one pandas data frame.

# Exercise 1

Combine the two data sets using the [pandas concat method](https://pandas.pydata.org/pandas-docs/stable/merging.html). In other words, find the union of the two data sets.

In [4]:
# TODO: import the pandas library
import pandas as pd

# TODO: read in each csv file into a separate variable
# HINT: remember from the Extract material that these csv file have some formatting issues
df_rural = pd.read_csv('rural_population_percent.csv', skiprows=4)
df_electricity = pd.read_csv('electricity_access_percent.csv', skiprows=4)

# TODO: remove the 'Unnamed:62' column from each data set
df_rural.drop('Unnamed: 62', axis=1, inplace=True)
df_electricity.drop('Unnamed: 62', axis=1, inplace=True)

# TODO: combine the two data sets together using the concat method
df = pd.concat([df_rural, df_electricity])

In [10]:
df.describe()

1960        1961        1962        1963        1964        1965  \
count  259.000000  259.000000  259.000000  259.000000  259.000000  259.000000   
mean    63.780723   63.268079   62.744129   62.210687   61.665423   61.126631   
std     23.872073   23.915329   23.962966   24.009611   24.050453   24.097359   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%     49.172500   48.369500   47.476500   46.669500   45.993000   45.333500   
50%     67.599000   66.478000   66.045000   65.338000   64.803000   64.470390   
75%     83.172084   82.764152   82.083313   81.629136   81.131221   80.953000   
max     97.923000   97.885000   97.846000   97.807000   97.767000   97.706000   

             1966        1967        1968        1969  ...        2008  \
count  259.000000  259.000000  259.000000  259.000000  ...  522.000000   
mean    60.589304   60.054138   59.499864   58.942975  ...   60.968028   
std     24.149868   24.195656   24.245253   24.288816  ...   32.038104   
min      0.000000    0.000000    0.000000    0.000000  ...    0.000000   
25%     44.372500   43.554000   42.463000   41.518000  ...   32.661500   
50%     63.952000   63.204000   62.640000   61.926773  ...   62.683322   
75%     80.677000   80.618000   80.395000   79.902500  ...   95.498157   
max     97.605000   97.499000   97.389000   97.274000  ...  100.000000   

             2009        2010        2011        2012        2013        2014  \
count  522.000000  522.000000  522.000000  521.000000  521.000000  521.000000   
mean    61.092004   60.932429   61.327071   61.589904   61.686117   61.925063   
std     32.134992   32.261980   32.150370   32.269978   32.386413   32.404793   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%     33.163000   32.638141   34.596250   35.385993   34.225000   34.124225   
50%     63.079904   63.249248   63.353546   62.980388   62.726070   62.469724   
75%     96.702669   96.947249   97.336274   97.824120   98.385853   98.701241   
max    100.000000  100.000000  100.000000  100.000000  100.000000  100.000000   

             2015        2016        2017  
count  521.000000  521.000000  260.000000  
mean    62.106432   62.353005   40.751455  
std     32.502656   32.460620   23.000182  
min      0.000000    0.000000    0.000000  
25%     34.908000   35.202616   21.449750  
50%     62.237572   62.349000   40.621500  
75%     99.032677   99.426155   59.538250  
max    100.000000  100.000000   91.730000  

[8 rows x 58 columns]

In [6]:
df_rural.head(2)

Country Name Country Code                            Indicator Name  \
0        Aruba          ABW  Rural population (% of total population)   
1  Afghanistan          AFG  Rural population (% of total population)   

   Indicator Code    1960    1961    1962   1963    1964   1965  ...    2008  \
0  SP.RUR.TOTL.ZS  49.224  49.239  49.254  49.27  49.285  49.30  ...  56.217   
1  SP.RUR.TOTL.ZS  91.779  91.492  91.195  90.89  90.574  90.25  ...  76.054   

     2009    2010    2011    2012    2013    2014    2015    2016    2017  
0  56.579  56.941  57.302  57.636  57.942  58.221  58.472  58.696  58.893  
1  75.687  75.311  74.926  74.532  74.129  73.718  73.297  72.868  72.430  

[2 rows x 62 columns]

In [7]:
df_electricity.head(2)

Country Name Country Code                           Indicator Name  \
0        Aruba          ABW  Access to electricity (% of population)   
1  Afghanistan          AFG  Access to electricity (% of population)   

   Indicator Code  1960  1961  1962  1963  1964  1965  ...       2008  \
0  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...  93.086166   
1  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...  42.400000   

        2009       2010       2011       2012       2013       2014  \
0  93.354546  93.356292  93.942375  94.255814  94.578262  94.906723   
1  44.854885  42.700000  43.222019  69.100000  67.259552  89.500000   

        2015       2016  2017  
0  95.238182  95.570145   NaN  
1  71.500000  84.137138   NaN  

[2 rows x 62 columns]

# Exercise 2 (Challenge)

This exercise is more challenging. Combine the data in the csv file together to the output that look like the following:

|Country Name|Country Code|Year|Rural_Value|Electricity_Value|
|--|--|--|--|--|--|
|Aruba|ABW|1960|49.224|49.239|
... etc.

Order the results in the dataframe by country and then by year

Here are a few pandas methods that should be helpful:
* [melt](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html)
* [merge](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.DataFrame.merge.html)
* [sort_values](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)

HINT: You can use country name, country code, and the year as common keys between the data sets

In [57]:
# TODO: merge the data sets together according to the instructions. First, use the 
# melt method to change the formatting of each data frame so that it looks like this:
# Country Name, Country Code, Year, Rural Value
# Country Name, Country Code, Year, Electricity Value

df_rural_melt = pd.melt(df_rural,\
                        id_vars=['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code'],\
                       var_name = 'Year', value_name='Electricity_Value')
df_electricity_melt = pd.melt(df_electricity,\
                              id_vars=['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code'],\
                             var_name='Year', value_name='Rural_Value')

# TODO: drop any columns from the data frames that aren't needed
df_rural_melt.drop(['Indicator Name', 'Indicator Code'], axis=1, inplace=True)
df_electricity_melt.drop(['Indicator Name', 'Indicator Code'], axis=1, inplace=True)

# TODO: merge the data frames together based on their common columns
# in this case, the common columns are Country Name, Country Code, and Year
df_merge = df_rural_melt.merge(df_electricity_melt, how='outer',\
                               on=['Country Name', 'Country Code', 'Year'])

# TODO: sort the results by country and then by year
df_combined = df_merge.sort_values(['Country Name', 'Year'])
df_combined

Country Name Country Code  Year  Electricity_Value  Rural_Value
1      Afghanistan          AFG  1960             91.779          NaN
265    Afghanistan          AFG  1961             91.492          NaN
529    Afghanistan          AFG  1962             91.195          NaN
793    Afghanistan          AFG  1963             90.890          NaN
1057   Afghanistan          AFG  1964             90.574          NaN
1321   Afghanistan          AFG  1965             90.250          NaN
1585   Afghanistan          AFG  1966             89.915          NaN
1849   Afghanistan          AFG  1967             89.570          NaN
2113   Afghanistan          AFG  1968             89.214          NaN
2377   Afghanistan          AFG  1969             88.848          NaN
2641   Afghanistan          AFG  1970             88.471          NaN
2905   Afghanistan          AFG  1971             88.083          NaN
3169   Afghanistan          AFG  1972             87.684          NaN
3433   Afghanistan          AFG  1973             87.274          NaN
3697   Afghanistan          AFG  1974             86.851          NaN
3961   Afghanistan          AFG  1975             86.417          NaN
4225   Afghanistan          AFG  1976             85.971          NaN
4489   Afghanistan          AFG  1977             85.513          NaN
4753   Afghanistan          AFG  1978             85.042          NaN
5017   Afghanistan          AFG  1979             84.565          NaN
5281   Afghanistan          AFG  1980             84.319          NaN
5545   Afghanistan          AFG  1981             84.070          NaN
5809   Afghanistan          AFG  1982             83.818          NaN
6073   Afghanistan          AFG  1983             83.563          NaN
6337   Afghanistan          AFG  1984             83.304          NaN
6601   Afghanistan          AFG  1985             83.042          NaN
6865   Afghanistan          AFG  1986             82.777          NaN
7129   Afghanistan          AFG  1987             82.509          NaN
7393   Afghanistan          AFG  1988             82.237          NaN
7657   Afghanistan          AFG  1989             81.962          NaN
...            ...          ...   ...                ...          ...
7655      Zimbabwe          ZWE  1988             72.476          NaN
7919      Zimbabwe          ZWE  1989             71.750          NaN
8183      Zimbabwe          ZWE  1990             71.012    30.357832
8447      Zimbabwe          ZWE  1991             70.262    30.718805
8711      Zimbabwe          ZWE  1992             69.501    28.200000
8975      Zimbabwe          ZWE  1993             69.060    31.436678
9239      Zimbabwe          ZWE  1994             68.665    28.100000
9503      Zimbabwe          ZWE  1995             68.268    32.130077
9767      Zimbabwe          ZWE  1996             67.868    32.459949
10031     Zimbabwe          ZWE  1997             67.466    32.774529
10295     Zimbabwe          ZWE  1998             67.061    33.070751
10559     Zimbabwe          ZWE  1999             66.653    38.400000
10823     Zimbabwe          ZWE  2000             66.242    33.597420
11087     Zimbabwe          ZWE  2001             65.830    33.830868
11351     Zimbabwe          ZWE  2002             65.415    34.200000
11615     Zimbabwe          ZWE  2003             65.521    34.266785
11879     Zimbabwe          ZWE  2004             65.706    34.481380
12143     Zimbabwe          ZWE  2005             65.890    34.701820
12407     Zimbabwe          ZWE  2006             66.074    37.200000
12671     Zimbabwe          ZWE  2007             66.257    35.184479
12935     Zimbabwe          ZWE  2008             66.440    35.457310
13199     Zimbabwe          ZWE  2009             66.622    43.369082
13463     Zimbabwe          ZWE  2010             66.804    36.063057
13727     Zimbabwe          ZWE  2011             66.985    36.900000
13991     Zimbabwe          ZWE  2012             67.166    36.728878
14255     Zimbab